src/main/java/com/google/sps/servlets/EventServlet.java

```diff
@@ -75,6 +85,122 @@ public void doPut(HttpServletRequest request, HttpServletResponse response) thro
+         && !startTimeAsString.isEmpty()){// event has time set
+      com.google.api.services.calendar.model.Event newEvent = createGCalendarEvent(startTimeAsString, startDateAsString,
+          duration, location, description, title, fields);
+      if (newEvent == null){
+        response.setContentType("text/html");
+        response.getWriter().print("<script>alert(\"Please login first.\")</script>");
+        RequestDispatcher dispatcher = request.getRequestDispatcher("/add_event.html");
+        dispatcher.include(request, response);
+        return ;
+      }
+      idFromGCalendarEvent = newEvent.getId();
+    }
+    String idForDataStoreEvent = null;
+    if (idFromGCalendarEvent == null)
+      idForDataStoreEvent = UUID.randomUUID().toString();
+    else idForDataStoreEvent = idFromGCalendarEvent;
+    Event event = new Event(idForDataStoreEvent, title, description, 
+        tags,
+        formatDateRange(startDateAsString),
+        formatTimeRange(startTimeAsString),
+        duration,
+        location, 
+        parseLinks(request.getParameter("links")), 
+        fields,
+        current_user_id, participants_ids, new ArrayList<String>(), new ArrayList<String>());
+
+    EventStorage.addEvent(event);
+    response.sendRedirect("/index.html");
+  }
+
+  private com.google.api.services.calendar.model.Event createGCalendarEvent(String timeInString, String dateInString, long durationInMinutes, 
+    String location, String description, String title, Map<String, String> externalFields) throws IOException {
+    com.google.api.services.calendar.model.Event event = new com.google.api.services.calendar.model.Event();
+    event.setDescription(description).setSummary(title);
+    DateTime startDateTime = new DateTime(dateInString + "T" + timeInString + ":00Z"); // Set in UTC
+    DateTime endDateTime = new DateTime(startDateTime.getValue() + durationInMinutes * 60 * 1000);
+    event.setStart(new EventDateTime().setDateTime(startDateTime));
+    event.setEnd(new EventDateTime().setDateTime(endDateTime));
+
+    ExtendedProperties extendedProps = new ExtendedProperties();
+    extendedProps.setShared(externalFields);
+    event.setExtendedProperties(extendedProps);
+
+    String calendarId = "primary";
+    try {
+      Calendar service = Utils.loadCalendarClient();
+      System.out.println(service);
+      if (service == null){
+        return null;
+      }
+      service.events().insert(calendarId, event).execute();
+      return event;
+    } catch (GeneralSecurityException e) {
+      e.printStackTrace();
+    }
+    return null;
+  }
+
+  private List<String> parseLinks(String links) {
+    return Arrays.asList(links.split(","));
+  }
+
+  private String formatDateRange(String date) {  // yyyy-mm-dd -> yyyy.mm.dd-yyyy.mm.dd
+    if (date != null && !date.isEmpty()) {
+      date = date.replace("-", ".");
+      return date + "-" + date;
+    }
+    return "0000.01.01-9999.12.12";
```

Comment:
Why are these defaults required here? If the date is not given, we would end up creating events with a very wide range -- I think the creation should just fail in that case.